File: frontend/src/routes/Inventory/lib/stockCalc.test.ts
import { describe, it, expect } from 'vitest';
import {
  sessionsRemaining,
  stockStatus,
  needsOrdering,
  orderUnits,
  orderBoxes,
  sortStock,
} from './stockCalc';

describe('sessionsRemaining', () => {
  it('returns qty for 1:1 items (SAK, cartridge, saline)', () => {
    expect(sessionsRemaining('SAK-303', 12)).toBe(12);
    expect(sessionsRemaining('CAR-172-C', 6)).toBe(6);
    expect(sessionsRemaining('UK00000880', 10)).toBe(10);
  });

  it('divides by 2 for needles (2 per session)', () => {
    expect(sessionsRemaining('P00012326', 20)).toBe(10);
    expect(sessionsRemaining('P00012326', 5)).toBe(2);
  });

  it('multiplies by 10 for PAK (1 per 10 sessions)', () => {
    expect(sessionsRemaining('PAK-001', 2)).toBe(20);
    expect(sessionsRemaining('PAK-001', 0)).toBe(0);
  });

  it('returns null for hospital items', () => {
    expect(sessionsRemaining('heparin', 5)).toBeNull();
    expect(sessionsRemaining('epo', 3)).toBeNull();
  });

  it('returns null for unknown code', () => {
    expect(sessionsRemaining('UNKNOWN', 5)).toBeNull();
  });
});

describe('stockStatus', () => {
  it('returns red when sessions remaining < 8', () => {
    expect(stockStatus('SAK-303', 7)).toBe('red');
    expect(stockStatus('SAK-303', 0)).toBe('red');
  });

  it('returns amber when sessions remaining 8–15', () => {
    expect(stockStatus('SAK-303', 8)).toBe('amber');
    expect(stockStatus('SAK-303', 15)).toBe('amber');
  });

  it('returns green when sessions remaining >= 16', () => {
    expect(stockStatus('SAK-303', 16)).toBe('green');
    expect(stockStatus('SAK-303', 24)).toBe('green');
  });

  it('returns green for hospital items with stock > 1', () => {
    expect(stockStatus('heparin', 4)).toBe('green');
  });

  it('returns red for hospital items with 0 stock', () => {
    expect(stockStatus('heparin', 0)).toBe('red');
  });

  it('returns amber for hospital items with 1 unit', () => {
    expect(stockStatus('heparin', 1)).toBe('amber');
  });
});

describe('needsOrdering', () => {
  it('returns true when stock below target', () => {
    expect(needsOrdering('SAK-303', 10)).toBe(true);   // target = 24
  });

  it('returns false when stock at or above target', () => {
    expect(needsOrdering('SAK-303', 24)).toBe(false);
    expect(needsOrdering('SAK-303', 30)).toBe(false);
  });

  it('always returns false for hospital items', () => {
    expect(needsOrdering('heparin', 0)).toBe(false);
    expect(needsOrdering('epo', 0)).toBe(false);
  });
});

describe('orderBoxes', () => {
  it('calculates boxes needed for SAK (boxSize=2)', () => {
    // target=24 bags, have 10 → need 14 → ceil(14/2) = 7 boxes
    expect(orderBoxes('SAK-303', 10)).toBe(7);
  });

  it('calculates boxes needed for cartridges (boxSize=6)', () => {
    // target=24, have 6 → need 18 → ceil(18/6) = 3 boxes
    expect(orderBoxes('CAR-172-C', 6)).toBe(3);
  });

  it('returns 0 when stock meets or exceeds target', () => {
    expect(orderBoxes('SAK-303', 24)).toBe(0);
    expect(orderBoxes('SAK-303', 30)).toBe(0);
  });
});

describe('sortStock', () => {
  it('puts needs-ordering items first', () => {
    const entries = [
      { code: 'SAK-303', qty: 30 },         // green, no order
      { code: 'CAR-172-C', qty: 5 },        // red, needs order
    ];
    const sorted = sortStock(entries);
    expect(sorted[0].code).toBe('CAR-172-C');
  });

  it('sorts by priority within the same needs-ordering group', () => {
    const entries = [
      { code: 'PAK-001', qty: 0 },       // needs ordering, priority 4
      { code: 'SAK-303', qty: 0 },       // needs ordering, priority 1
    ];
    const sorted = sortStock(entries);
    expect(sorted[0].code).toBe('SAK-303');
  });

  it('places hospital items after all nxstage items', () => {
    const entries = [
      { code: 'heparin', qty: 0 },
      { code: 'SAK-303', qty: 0 },
    ];
    const sorted = sortStock(entries);
    expect(sorted[sorted.length - 1].code).toBe('heparin');
  });
});

File: flutter/test/stock_calc_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:home_hd/features/inventory/stock_calc.dart';

void main() {
  group('sessionsRemaining', () {
    test('1:1 items return qty', () {
      expect(sessionsRemaining('SAK-303', 12), 12);
      expect(sessionsRemaining('CAR-172-C', 6), 6);
      expect(sessionsRemaining('UK00000774', 24), 24); // On/Off Pack now 1/session
    });
    test('needles divide by 2, PAK multiplies by 10', () {
      expect(sessionsRemaining('P00012326', 20), 10);
      expect(sessionsRemaining('PAK-001', 2), 20);
    });
    test('hospital + unknown return null', () {
      expect(sessionsRemaining('heparin', 5), isNull);
      expect(sessionsRemaining('UNKNOWN', 5), isNull);
    });
  });

  group('stockStatus', () {
    test('nxstage thresholds (red <8, amber 8-15, green >=16)', () {
      expect(stockStatus('SAK-303', 7), StockStatus.red);
      expect(stockStatus('SAK-303', 8), StockStatus.amber);
      expect(stockStatus('SAK-303', 16), StockStatus.green);
    });
    test('hospital: 0 red, 1 amber, >1 green', () {
      expect(stockStatus('heparin', 0), StockStatus.red);
      expect(stockStatus('heparin', 1), StockStatus.amber);
      expect(stockStatus('heparin', 4), StockStatus.green);
    });
  });

  group('consumedUnits', () {
    test('1:1 session items scale linearly', () {
      expect(consumedUnits('SAK-303', 20), 20);
      expect(consumedUnits('CAR-172-C', 16), 16);
    });
    test('needles are 2/session', () {
      expect(consumedUnits('P00012326', 20), 40);
    });
    test('PAK is 1 per 10 sessions, ceil', () {
      expect(consumedUnits('PAK-001', 20), 2);
      expect(consumedUnits('PAK-001', 15), 2); // ceil(15/10)
    });
    test('monthly items return 0 (no per-session rate)', () {
      expect(consumedUnits('UK00000830', 20), 0); // Sani-Cloth
      expect(consumedUnits('UK00000832', 20), 0); // Sharps Bin
    });
  });

  group('orderBoxes', () {
    test('falls back to fixed targetQty when no delivery date given', () {
      expect(orderBoxes('SAK-303', 10), 7); // ceil((24-10)/2)
      expect(orderBoxes('CAR-172-C', 6), 3); // ceil((24-6)/6)
      expect(orderBoxes('SAK-303', 24), 0);
    });

    test('excludes backup qty from working stock', () {
      // working = 10-8 = 2, target = 24, order = ceil((24-2)/2) = 11
      expect(orderBoxes('SAK-303', 10, backupQty: 8), 11);
      expect(orderBoxes('SAK-303', 10, backupQty: 0), 7);
    });

    test('delivery sessions: normal order (7-day lead = 4 sessions + 16 buffer = 20)', () {
      // Need 20 bags, have 10 working → order 10 → ceil(10/2) = 5 boxes
      expect(orderBoxes('SAK-303', 10, deliverySessions: 20), 5);
    });

    test('delivery sessions: early order (3-week lead = 12 sessions + 16 buffer = 28)', () {
      // Need 28 bags, have 10 working → order 18 → ceil(18/2) = 9 boxes
      expect(orderBoxes('SAK-303', 10, deliverySessions: 28), 9);
    });

    test('delivery sessions with backup: excludes backup from working stock', () {
      // 20 total, 8 backup → working = 12; need 28; order 16 → ceil(16/2) = 8 boxes
      expect(orderBoxes('SAK-303', 20, backupQty: 8, deliverySessions: 28), 8);
    });

    test('monthly items fall back to targetQty even with delivery sessions', () {
      // Sani-Cloth has no per-session rate; consumedUnits returns 0 → use targetQty=1
      expect(orderBoxes('UK00000830', 0, deliverySessions: 28), 1);
      expect(orderBoxes('UK00000830', 1, deliverySessions: 28), 0);
    });
  });

  test('needsOrdering false for hospital, true below target', () {
    expect(needsOrdering('SAK-303', 10), isTrue);
    expect(needsOrdering('SAK-303', 24), isFalse);
    expect(needsOrdering('heparin', 0), isFalse);
  });

  test('sortStock: needs-ordering first, then priority, hospital after nxstage', () {
    final sorted = sortStock(const [
      StockEntry('heparin', 0),
      StockEntry('SAK-303', 30), // green, no order
      StockEntry('CAR-172-C', 5), // needs order
    ]);
    expect(sorted.first.code, 'CAR-172-C'); // needs ordering first
    expect(sorted.last.code, 'heparin'); // hospital after nxstage
  });
}
